### PR TITLE
Cli surface proposal

### PR DESCRIPTION
**Pull request by loosebazooka**
Proposal for design for the cli surface. Implementation will follow

Rendered: https://github.com/GoogleContainerTools/jib/blob/cli-surface-proposal/proposals/jib-cli-surface.md

### CODE REVIEW

File: proposals/jib-cli-surface.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
+# Jib CLI Surface Specification
+
+Specification for a command structures exposed to users of the jib-cli
+
+## Usage
+`jib command [OPTIONS]`
+
+## Options
```
Comment (mpeddada1) [inline proposals/jib-cli-surface.md:8]: super nit: maybe switch the ordering of the "Commands" section and "Options" section to match the format in the usage section

File: proposals/jib-cli-surface.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+```
+-t  --target <image-ref>  a target image reference
+```
+
+### Optional
+```
+-b  --build-file <file>                    location of the build file (default ./jib.yaml)
+    --credHelper <registry>=<credHelper>   credential helper to use for a registry (repeatable)
+    --docker-config <directory>            location of docker configuration
+    --help                                 print usage
```
Comment (mpeddada1) [inline proposals/jib-cli-surface.md:20]: Are these indentations intended?

Comment (loosebazooka) [reply proposals/jib-cli-surface.md:20]: yeah, generally following the format of some command `--help` (like `ls --help`) although I think `java --help` has a slightly different style. Anyway, the indentations are intentional, I can clean it up a little more though

File: proposals/jib-cli-surface.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+```
+-b, --build-file <file>                    location of the build file (default <context>/jib.yaml)
+-c, --context <dir>                        location of the build context (default .)
+    --credHelper <registry>=<credHelper>   credential helper to use for a registry (repeatable)
+    --docker-config <directory>            location of docker configuration
+    --help                                 print usage
+-p, --parameter <name>=<value>             templating parameters replace `${name}` with `value` in the build file (repeatable)
+    --tag <tag1>[,<tag2>,...]              additional tags for target
+    --verbosity <level>                    set logging verbosity (error, warn, lifecycle (default), info, debug)
+-v, --version                              print version information and exit
```
Comment (mpeddada1) [inline proposals/jib-cli-surface.md:34]: Is this the general jib-cli version?

Comment (mpeddada1) [inline proposals/jib-cli-surface.md:34]: Is this the general jib-cli version?

Comment (loosebazooka) [reply proposals/jib-cli-surface.md:34]: yeah this would be whatever version information we bake into the system.

Comment (mpeddada1) [reply proposals/jib-cli-surface.md:34]: nice. we could add this maybe?
```suggestion
-v, --version                              print jib-cli version information and exit 
```

File: proposals/jib-cli-surface.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+## Options
+
+### Required
+```
+-t  --target <image-ref>  a target image reference
+```
+
+### Optional
+```
```
Comment (mpeddada1) [inline proposals/jib-cli-surface.md:25]: This is just a thought and also very much optional: How about having  a separate section for "general purpose flags" such as `help`, `version` and `verbosity`?

File: proposals/jib-cli-surface.md (hunk 3 above)
Comment (loosebazooka) [reply proposals/jib-cli-surface.md:34]: I think when you run `tool --version`, it's expected that you are expecting the version of the tool, no?

File: proposals/jib-cli-surface.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+## Usage
+`jib COMMAND [OPTIONS]`
+
+## Commands
+
+The jib cli exposes the following top level commands, mirroring what is available in the plugins
+```
+build        build to registry
+dockerBuild  build to docker daemon
+buildTar     build to a tar file
```
Comment (chanseokoh) [inline proposals/jib-cli-surface.md:14]: Just food for thoughts:

We have had multiple instances of feedback to be able to "warm up a cache" without pushing or creating a tar. Honestly, I wish the Jib build plugins had a way. And if we support the URI style, I think there's no need to have three commands.

Related, we may considering providing an option to configure the application layer cache (the default is to use a temporary directory) and the base image cache, but I'm not sure how useful they will be in practice.

File: proposals/jib-cli-surface.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+### Required
+```
+-t  --target <image-ref>  a target image reference
+```
+
+### Optional
+```
+-b, --build-file <file>                    location of the build file (default <context>/jib.yaml)
+-c, --context <dir>                        location of the build context (default .)
+    --credHelper <registry>=<credHelper>   credential helper to use for a registry (repeatable)
```
Comment (chanseokoh) [inline proposals/jib-cli-surface.md:28]: No simple way to pass a username and password pair if there's no credential helper or a Docker CLI? However, I don't know what's a conventional way to safely pass credentials for a CLI tool.

Anyways, `--allow-insecure` or `--insecure-registries=gcr.io,ghcr.io,...` sounds like a useful option. (Unfortunately, jib-core doesn't support per-registry configuration.)

File: proposals/jib-cli-surface.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+-p, --parameter <name>=<value>             templating parameters replace `${name}` with `value` in the build file (repeatable)
+    --tag <tag1>[,<tag2>,...]              additional tags for target
+```
+
+#### Auth/Security
+```
+    --allow-insecure-registries            allow jib to communicate with registries over https
+    --credHelper <registry>=<credHelper>   credential helper to use for a registry (repeatable)
+    --send-credentials-over-http           allow jib to send credentials over http (used in conjunction with --allow-insecure-registries)
+    --to-auth <username> <password>        configure a username and password for authentication on the registry that an image is being built to
```
Comment (loosebazooka) [inline proposals/jib-cli-surface.md:45]: I think maybe I might split this into
```
--to-auth-username <username>
--to-auth-password <password>
```
this lets us take advantage of some password related tools that pico cli adds

File: proposals/jib-cli-surface.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    --docker-config <directory>            location of docker configuration
+    --name <image-ref>                     image name to bake into tar file (only available when using -t tar://...)
+-p, --parameter <name>=<value>             templating parameters replace `${name}` with `value` in the build file (repeatable)
+    --tag <tag1>[,<tag2>,...]              additional tags for target
+```
+
+#### Auth/Security
+```
+    --allow-insecure-registries            allow jib to communicate with registries over https
+    --credHelper <registry>=<credHelper>   credential helper to use for a registry (repeatable)
```
Comment (loosebazooka) [inline proposals/jib-cli-surface.md:43]: maybe this should be --to-credHelper

File: proposals/jib-cli-surface.md [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+#### Auth/Security
+```
+    --allow-insecure-registries            allow jib to communicate with registries over https
+    --send-credentials-over-http           allow jib to send credentials over http (used in conjunction with --allow-insecure-registries)
+```
+##### Credentials
+
+Credentials can be specified using credential helpers or username + password. The following options are available
+```
+    --credHelper <credHelper>              credential helper to use for a registry
```
Comment (loosebazooka) [inline proposals/jib-cli-surface.md:51]: this is all wrong, credential helpers already encode the server url, we do no need to,from,etc for cred helpers

Comment (chanseokoh) [reply proposals/jib-cli-surface.md]: Some part of the general Docker auth has bothered me a bit. For example, whether using a credential helper or `docker login`, it didn't seem like you can provide different usernames for the base and the target repos if the repositories are on the same registry. That is, say you have two private Docker Hub repos `alice/image` and `bob/image` (i.e., repos in two different accounts), then you have to repeatedly do `docker login -u alice` and `docker login -u bob` if you need to work on both repos. Using a credential helper doesn't help either; FWIW, credentials are scoped solely by the registry.

Maybe it's just rare in practice that one has two different login accounts on the same registry? Anyways, I think having a single credential helper makes sense. Theoretically feasible, but I think it's unlikely that someone has multiple credential helpers set up only to return different accounts for the same registry.

File: proposals/jib-cli-surface.md (hunk 5 above)
Comment (loosebazooka) [reply proposals/jib-cli-surface.md]: Added a warm up cache section at the end, presumably that would be a separate command? we'd have to provide a shortcut in jib-core stepsrunner to do it though

File: proposals/jib-cli-surface.md (hunk 6 above)
Comment (loosebazooka) [reply proposals/jib-cli-surface.md]: Added some more flags here to cover these

File: proposals/jib-cli-surface.md (hunk 9 above)
Comment (loosebazooka) [reply proposals/jib-cli-surface.md]: Yeah, I can see that, we can always add in to/from later. It seems strange for now to do it though.

File: proposals/jib-cli-surface.md [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                          tar:///aboslute/path.tar
+```
+
+### Optional
+
+
+#### Build Config
+```
+    --application-cache <directory>        location of the application cache
+    --base-image-cache <directory>         location of the base image cache
```
Comment (loosebazooka) [inline proposals/jib-cli-surface.md:33]: define default?

File: proposals/jib-cli-surface.md [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+-b, --build-file <file>                    location of the build file (default <context>/jib.yaml)
+-c, --context <dir>                        location of the build context (default .)
+    --docker-config <directory>            location of docker configuration
+    --name <image-ref>                     image name to bake into tar file (only available when using -t tar://...)
```
Comment (loosebazooka) [inline proposals/jib-cli-surface.md:37]: is this required or optional?

File: proposals/jib-cli-surface.md [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+1. `--username`, `--password`
+1. `--to-username`, `--to-password`
+1. `--from-username`, `--from-password`
+1. `--to-username`, `--to-password`, `--from-username`, `--from-password`
+
+
+#### Info Params
+```
+    --help                                 print usage and exit
+    --stacktrace                           print stacktrace on error (for debugging issues in the jib-cli)
```
Comment (loosebazooka) [inline proposals/jib-cli-surface.md:70]: should be hidden in help?

Comment (loosebazooka) [reply proposals/jib-cli-surface.md:70]: also enable http tracing? (default level = config, output = console)

File: proposals/jib-cli-surface.md [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+    --verbosity <level>                    set logging verbosity (error, warn, lifecycle (default), info, debug)
+-v, --version                              print version information and exit 
+```
```
Comment (loosebazooka) [inline proposals/jib-cli-surface.md:73]: --serialize also hidden

File: proposals/jib-cli-surface.md [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+## Environment Variables
+
+### Docker daemon specification
+
+Information on communicating with the docker daemon is determined using the `DOCKER_*` environment variables.
+
+## Other considerations
+
+- Users have asked to warm up the cache, perhaps we can have a cache build?
```
Comment (loosebazooka) [inline proposals/jib-cli-surface.md:87]: jib core needs to support this?

File: proposals/jib-cli-surface.md (hunk 11 above)
Comment (loosebazooka) [reply proposals/jib-cli-surface.md:37]: Name is required for building "tar" images (https://github.com/GoogleContainerTools/jib/blob/master/jib-core/src/main/java/com/google/cloud/tools/jib/api/Containerizer.java#L104)

File: proposals/jib-cli-surface.md (hunk 10 above)
Comment (loosebazooka) [reply proposals/jib-cli-surface.md]: I think there's some work to be done here to define a better default, but the default that jib-core uses is a temporary directory.

Comment (loosebazooka) [reply proposals/jib-cli-surface.md]: added to "Other Considerations" section

File: proposals/jib-cli-surface.md (hunk 11 above)
Comment (loosebazooka) [reply proposals/jib-cli-surface.md]: updated text

File: proposals/jib-cli-surface.md [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    --send-credentials-over-http           allow jib to send credentials over http (used in conjunction with --allow-insecure-registries)
+```
+##### Credentials
+
+Credentials can be specified using credential helpers or username + password. The following options are available
+```
+    --credential-helper <credHelper>       credential helper to use for a registry, a path or name suffix (docker-credential-<suffix>) (repeatable)
+
+    --username <username>                  configure a username for authenticating against registries
+    --password <password>                  configure a passowrd for authenticating against registries (interactive if <password> is omitted)
```
Comment (mpeddada1) [inline proposals/jib-cli-surface.md:54]:
```suggestion
    --password <password>                  configure a password for authenticating against registries (interactive if <password> is omitted)
```

**Review by mpeddada1 — APPROVED** [review]

**Pull request merged**